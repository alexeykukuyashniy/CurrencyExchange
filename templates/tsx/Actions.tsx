import {
    CANCEL_EDIT,
    EDIT_BUY,
    EDIT_BUY_STEP2,
    EDIT_SELL,
    EDIT_SELL_STEP2,
    EDIT_TRANSFER,
    RATE_UPDATED,
    SAVE_EDIT,
    TOKEN_SET, VIEW_HOME
} from "./Constants";

export function editBuy(currencyid:number,step:number){
    const action = {
        type: (step == 1? EDIT_BUY : EDIT_BUY_STEP2)/*,
        currencyid,
        step*/
    }
    return action;
}

export function editSell(currencyid:number,step:number){
    const action = {
        type: (step == 1? EDIT_SELL : EDIT_SELL_STEP2)/*,
        currencyid,
        step*/
    }
    return action;
}

export function cancelEdit(){
    const action = {
        type:CANCEL_EDIT
    }
    return action;
}

export function saveEdit(){
    const action = {
        type: SAVE_EDIT
    }
    return action;
}

export function view(){
    const action = {
        type: VIEW_HOME
    }
    return action;
}

export function editTransfer(){
    const action = {
        type: EDIT_TRANSFER
    }
    return action;
}

export function updateRate(data:any){
    const action = {
        type: RATE_UPDATED,
        data: data
    }
    return action;
}

export function setToken(data:any){
    const action = {
        type: TOKEN_SET,
        token: data
    }
    return action;
}