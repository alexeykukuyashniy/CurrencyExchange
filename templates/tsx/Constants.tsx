
// db settings
export const COMMISSION = "Commission";
export const SURCHARGE = "Surcharge";
export const REFRESH_PERIOD = "RefreshPeriod";
export const MINIMAL_COMMISSION = "MinimalCommission";
export const BUY_SELL_RATE_MARGIN = "BuySellRateMargin";
export const MINIMAL_CURRENCY_REST = "MinimalCurrencyRest";

// redux actions & states
export const VIEW_HOME = "VIEW_HOME";
export const EDIT_BUY = "EDIT_BUY";
export const EDIT_BUY_STEP2 = "EDIT_BUY_STEP2";
export const EDIT_SELL = "EDIT_SELL";
export const EDIT_SELL_STEP2 = "EDIT_SELL_STEP2";
export const CANCEL_EDIT = "CANCEL_EDIT";
export const SAVE_EDIT = "SAVE_EDIT";
export const EDIT_TRANSFER = "EDIT_TRANSFER";
export const RATE_UPDATED = "RATE_UPDATED";
export const TOKEN_SET = "TOKEN_SET";
export const NO_TOKEN = "NO_TOKEN";
export const SETTINGS_UPDATED = "SETTINGS_UPDATED";

export const enum TR_TRANSACTION_TYPES {BUY = 1, SELL = 2, SEND = 3, RECEIVE = 4}

export interface IStoreState {
    state: string;
    data: any|undefined;
}

export interface ISecurityStoreState {
    state: string;
    token: string|undefined;
}

export interface IRate {
    currencyid: number;
    code: string;
    namee: string|undefined;
    buyrate: number;
    sellrate: number;
    date: string;
    amount: number|undefined;
}

export interface ISettings {
    refreshPeriod: string;
    commission: string;
    surcharge: string;
    minimalCommission: string;
    buySellRateMargin: string;
}
