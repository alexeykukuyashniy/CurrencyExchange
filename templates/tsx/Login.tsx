import * as React from "react";
import axios from "axios";
import store from "./Store";
import {setToken} from "./Actions";

export class CELogin extends React.Component<{}, {user: string, pwd: string, status: string}> {

    constructor(props: any) {
        super(props);

        this.login = this.login.bind(this);
        this.handleInputChange = this.handleInputChange.bind(this);

        this.state = {
            pwd: "",
            status: "",
            user: ""
        };
    }

    public render() {
        return (<div id="dvLogin">
                <form method="POST">
                    <table>
                        <tbody>
                        <tr>
                            <td/>
                            <td>
                                <h5>
                                    Enter Login
                                </h5>
                            </td>
                        </tr>
                        <tr>
                            <td/>
                            <td>
                                <span id="spStatus" style={{marginLeft: "1em", color: "red"}}>{this.state.status}</span>
                            </td>
                        </tr>
                        <tr>
                            <td>
                                User:
                            </td>
                            <td>
                                <input
                                    id="user"
                                    name="user"
                                    type="text"
                                    value={this.state.user}
                                    onChange={this.handleInputChange}
                                />
                            </td>
                        </tr>
                        <tr>
                            <td>
                                Password:
                            </td>
                            <td>
                                <input
                                    name="pwd"
                                    type="text"
                                    value={this.state.pwd}
                                    onChange={this.handleInputChange}
                                />
                            </td>
                        </tr>
                        <tr>
                            <td colSpan={2} style={{textAlign: "right"}}>
                                <input type="button" onClick={this.login} value="Login" disabled={!this.isDataValid()}/>
                            </td>
                        </tr>
                        </tbody>
                    </table>
                </form>
            </div>
        );
    }

    private isDataValid() {
        return this.state.user.length > 0 &&
               this.state.pwd.length > 0;
    }

    private handleInputChange(event: any) {
        const name: string = event.target.attributes.getNamedItem("name").value;
        const val: string = event.target.value;
        if (name === "user") {
            this.setState({user: val});
        } else {
            this.setState({pwd: val});
        }
    }

    private login(event: any) {
        const that = this;
        const data = {
            pwd: this.state.pwd,
            user: this.state.user
        };

        axios.post("/doLogin", data).then((response) => {
            if (response.data === "Incorrect password") {
                that.setState({status: response.data});
            } else {
                that.setState({status: ""});
                store.dispatch(setToken(response.data));
            }
        }).catch((error) => {
            console.log(error);
        });
    }
}

export default CELogin;
