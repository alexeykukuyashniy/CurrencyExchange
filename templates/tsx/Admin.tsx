import * as React from "react";
import axios from "axios";
import * as constants from "./Constants";
import store, {StoreUtils} from "./Store";
import {view, updateSettings} from "./Actions";

interface ISetting {
  name: string;
  value: string;
}

export class Admin extends React.Component<{}, constants.ISettings> {

    public loaded: boolean = false;
    private status: string = "";

    constructor(props: any) {
        super(props);

        this.handleInputChange = this.handleInputChange.bind(this);
        this.fetchData = this.fetchData.bind(this);
        this.isDataValid = this.isDataValid.bind(this);
        this.saveSettings = this.saveSettings.bind(this);
        this.clearStatus = this.clearStatus.bind(this);

        this.state = {
            buySellRateMargin: "",
            commission: "",
            minimalCommission: "",
            refreshPeriod: "",
            surcharge: ""
        };
        store.dispatch(view()); // in case edit form is active
    }

    public render() {
        console.log("rendering...");
        if (!this.loaded) {
            return <div id="dvSettings">Loading...</div>;
        }
        return (
            <div id="dvSettings">
                <form method="POST">
                    <h5>
                        Change setting below and click Update
                    </h5>
                    <div className="horizontal-rule"/>
                    <table className="tblSetting">
                        <tbody>
                        <tr>
                            <td>
                                Refresh currency exchange rates every&nbsp;
                            </td>
                            <td>
                                <input
                                    name={constants.REFRESH_PERIOD}
                                    type="text"
                                    value={this.state.refreshPeriod}
                                    onChange={this.handleInputChange}
                                />
                                &nbsp;seconds (0-60)
                            </td>
                        </tr>
                        <tr>
                            <td>
                                Commission, % (0-3):
                            </td>
                            <td>
                                <input
                                    name={constants.COMMISSION}
                                    type="text"
                                    value={this.state.commission}
                                    onChange={this.handleInputChange}
                                />
                            </td>
                        </tr>
                        <tr>
                            <td>
                                Surcharge, USD (0-3):
                            </td>
                            <td>
                                <input
                                    name={constants.SURCHARGE}
                                    type="text"
                                    value={this.state.surcharge}
                                    onChange={this.handleInputChange}
                                />
                            </td>
                        </tr>
                        <tr>
                            <td>
                                Minimal Commission, USD (0-3):
                            </td>
                            <td>
                                <input
                                    name={constants.MINIMAL_COMMISSION}
                                    type="text"
                                    value={this.state.minimalCommission}
                                    onChange={this.handleInputChange}
                                />
                            </td>
                        </tr>
                        <tr>
                            <td>
                                Buy/Sell rate margin, % (0-2):
                            </td>
                            <td>
                                <input
                                    name={constants.BUY_SELL_RATE_MARGIN}
                                    type="text"
                                    value={this.state.buySellRateMargin}
                                    onChange={this.handleInputChange}
                                />
                            </td>
                        </tr>
                        </tbody>
                    </table>
                    <div className="horizontal-rule"/>
                    <input type="button" onClick={this.saveSettings} value="Update" disabled={!this.isDataValid()}/>
                    <span id="spStatus" style={{marginLeft: "1em", fontStyle: "italic"}}>{this.status}</span>
                </form>
            </div>
        );
    }

    public fetchData() {
        try {
            const that = this;
            fetch("./settings", StoreUtils.authHeader()).then((response) => {
                if (response.ok) {
                    let commission: string = "";
                    let refreshPeriod: string = "";
                    let surcharge: string = "";
                    let minimalCommission: string = "";
                    let buySellRateMargin: string = "";
                    const data = response.json();
                    data.then((d) => {
                        const settings = (d as ISetting[]);
                        for (const s of settings) {
                            switch (s.name) {
                                case constants.REFRESH_PERIOD:
                                    refreshPeriod = s.value;
                                    break;
                                case constants.COMMISSION:
                                    commission = s.value;
                                    break;
                                case constants.SURCHARGE:
                                    surcharge = s.value;
                                    break;
                                case constants.MINIMAL_COMMISSION:
                                    minimalCommission = s.value;
                                    break;
                                case constants.BUY_SELL_RATE_MARGIN:
                                    buySellRateMargin = s.value;
                                    break;
                            }
                        }

                        that.setState({
                            buySellRateMargin,
                            commission,
                            minimalCommission,
                            refreshPeriod,
                            surcharge
                        });
                        that.loaded = true;
                        that.forceUpdate();
                    });
                }
            });
        } catch (e) {
            console.log(e);
        }
    }

    public componentDidMount() {
        this.fetchData();
    }

    private handleInputChange(event: any) {
        console.log(event.target.value);
        console.log(event.target.attributes.getNamedItem("name").value);
        console.log(event.target);
        const name: string = event.target.attributes.getNamedItem("name").value;
        const updState: constants.ISettings = this.state;
        let val: string = event.target.value;

        if (val.length > 0) {
            console.log(val.indexOf("."));
            if (name === constants.REFRESH_PERIOD /*&& parseInt(val) == NaN*/) {
                val = parseInt(val, 10).toString();
            } else if (val.indexOf(".") > 0 && val[val.length - 1] !== "." && val.indexOf(".") <  val.length - 3) {
                val = parseFloat(val).toString();
                // leave 2 digits after dot
                const val2: string = Number(Math.floor(parseFloat(val) * 100 ) / 100).toFixed(2) ;
                if (val !== val2) {
                    val = val2;
                }
            }
        }

        if (val === "NaN") {
            return;
        }
        console.log("set value", val);
        switch (name) {
            case constants.COMMISSION :
                updState.commission = val;
                break;
            case constants.REFRESH_PERIOD :
                updState.refreshPeriod = val;
                break;
            case constants.SURCHARGE :
                updState.surcharge = val;
                break;
            case constants.MINIMAL_COMMISSION :
                updState.minimalCommission = val;
                break;
            case constants.BUY_SELL_RATE_MARGIN :
                updState.buySellRateMargin = val;
                break;
        }
        this.setState(updState, this.clearStatus);
    }

    private clearStatus() {
        this.status = "";
    }

    private isDataValid() {
        return this.state.refreshPeriod.length > 0 &&
            this.state.commission.length > 0 &&
            this.state.surcharge.length > 0 &&
            this.state.minimalCommission.length > 0 &&
            this.state.buySellRateMargin.length > 0 &&
            parseInt(this.state.refreshPeriod, 10) >= 0 &&
            parseInt(this.state.refreshPeriod, 10) <= 60 &&
            parseFloat(this.state.commission) >= 0 &&
            parseFloat(this.state.commission) <= 3 &&
            parseFloat(this.state.surcharge) >= 0 &&
            parseFloat(this.state.surcharge) <= 3 &&
            parseFloat(this.state.minimalCommission) >= 0 &&
            parseFloat(this.state.minimalCommission) <= 3 &&
            parseFloat(this.state.buySellRateMargin) >= 0 &&
            parseFloat(this.state.buySellRateMargin) <= 2;
    }

    private saveSettings(event: any) {
        const that = this;
        const data: constants.ISettings = {
                buySellRateMargin: this.state.buySellRateMargin,
                commission: this.state.commission,
                minimalCommission: this.state.minimalCommission,
                refreshPeriod: this.state.refreshPeriod,
                surcharge: this.state.surcharge
            };

        const dataSave = {
            BuySellRateMargin: this.state.buySellRateMargin,
            Commission: this.state.commission,
            MinimalCommission: this.state.minimalCommission,
            RefreshPeriod: this.state.refreshPeriod,
            Surcharge: this.state.surcharge
        };
        axios.post("/savesettings", dataSave, StoreUtils.authHeader()).then((response) => {
                that.status = "Saved";
                store.dispatch(updateSettings(data));
                that.forceUpdate();
            })
            .catch((error) => {
                console.log(error);
            });
    }
}

export default Admin;
