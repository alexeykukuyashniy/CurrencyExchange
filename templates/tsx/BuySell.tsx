import * as React from 'react';
import 'primereact/resources/themes/nova-light/theme.css';
import 'primereact/resources/primereact.min.css';
import 'primeicons/primeicons.css';
import { SubmissionError } from 'redux-form';
import { Field, reduxForm } from 'redux-form'
import * as constants from "./Constants";
import store, {StoreUtils} from './Store';
import {cancelEdit, editBuy, editSell, saveEdit} from "./Actions";
import axios from 'axios';

interface IFormData{
    amount:number;
}

interface ISettings {
    commission: number;
    surcharge: number;
    minimalCommission: number;
    buySellRateMargin:number;
}

interface setting {
    name: string;
    value: string;
}

interface BuySellProps{
    rate:number;
    amount:number;
    currencycode:string;
    currencyid:number;
}

interface BuySellState{
    rate:number;
    amount:number;
    amountRest:number;
    currencycode:string;
    currencyid:number;
    subtotal:number;
    total:number;
    commissionamount:number;
    settings:ISettings;
    step:number|undefined; // needed to refresh the form on back/next button click
}

   class BuySell extends React.Component<BuySellProps, BuySellState> {

       formData:IFormData = {amount:0};
       unsubscribe:any;

       constructor(props: BuySellProps) {
           super(props);
           this.amountChange = this.amountChange.bind(this);
           this.calcTotals = this.calcTotals.bind(this);
           this.submit = this.submit.bind(this);
           this.fetchSettings = this.fetchSettings.bind(this);
           this.cancelClick = this.cancelClick.bind(this);
           this.backClick = this.backClick.bind(this);
           this.stepAction = this.stepAction.bind(this);
           this.handleStateChange = this.handleStateChange.bind(this);
           this.getAmountRest = this.getAmountRest.bind(this);

           this.state = {
               rate: props.rate,
               amount: props.amount,
               amountRest: 0,
               currencycode: props.currencycode,
               currencyid: props.currencyid,
               commissionamount:0,
               subtotal:0,
               total:0,
               step:0,
               settings : {commission: 0, surcharge: 0, minimalCommission: 0, buySellRateMargin: 0}
           };
           this.unsubscribe = store.subscribe(this.handleStateChange);
           this.fetchSettings();
           this.getAmountRest(props.currencycode);
       }

       getAmountRest(code:string) {
           let that = this;
           fetch('./cashamount?code=' + (this.isSell() ? code : 'USD'), StoreUtils.authHeader()).then(function (response) {
               if (response.ok) {

                   let data = response.json();
                   data.then(data => {
                          console.log(data, (data as {amount:string}[])[0]);
                          let amount:number = parseFloat((data as {amount:string}[])[0].amount.replace(",",""));
                          that.setState({amountRest:amount});
                          console.log('amount rest:', amount);
                       }
                   );
               }
           });
       }

       componentWillUnmount(): void {
           console.log('componentWillUnmount');
           this.unsubscribe();
       }

       handleStateChange() {
           console.log('BS handleStateChange: ', StoreUtils.getStoreState(), store.getState());
           console.log(store.getState().main);
           console.log(store.getState().main.data);
           console.log(store.getState().main.data.data);

           if (StoreUtils.getStoreState() == constants.EDIT_BUY_STEP2 || StoreUtils.getStoreState() == constants.EDIT_SELL_STEP2)
           {
               return;
           }

           if (store.getState().main == undefined || store.getState().main.data == undefined || store.getState().main.data.data == undefined) {
               console.log('no data. exiting 1');
               return;
           }
           let rates = store.getState().main.data.data as constants.IRate[];
           if (rates == undefined) {
               console.log('no data. exiting 2');
               return;
           }
           console.log('rates: ', rates);
           let currencyid = this.state.currencyid;
           let rate = 0;
           for (let i: number = 0; i < rates.length; i++) {

               if (currencyid == rates[i].currencyid) {
                   rate = (StoreUtils.getStoreState() == constants.EDIT_BUY || StoreUtils.getStoreState() == constants.EDIT_BUY_STEP2 ? rates[i].buyrate : rates[i].sellrate);
               }
           }

           console.log('found rate:', rate);
           try {
               this.setState({rate: rate}, this.calcTotals);
           } catch(err)
           {
               console.log('ERROR: ', err);
           }
       }

       fetchSettings() {
           let that = this;
           fetch('./settings', StoreUtils.authHeader()).then(function (response) {
               if (response.ok) {
                   let commission: number = 0;
                   let surcharge: number = 0;
                   let minimalCommission: number = 0;
                   let buySellRateMargin: number = 0;
                   let data = response.json();
                   data.then(data => {
                       let settings = (data as setting[]);
                       for (let i = 0; i < settings.length; i++) {
                           console.log(settings[i]);

                           switch (settings[i].name) {
                               case constants.COMMISSION:
                                   commission = settings[i].value as unknown as number;
                                   break;
                               case constants.SURCHARGE:
                                   surcharge = settings[i].value as unknown as number;
                                   break;
                               case constants.MINIMAL_COMMISSION:
                                   minimalCommission = settings[i].value as unknown as number;
                                   break;
                               case constants.BUY_SELL_RATE_MARGIN:
                                   buySellRateMargin = settings[i].value as unknown as number;
                                   break;
                           }
                       }

                       that.setState({
                               settings: {
                               commission: commission,
                               surcharge: surcharge,
                               minimalCommission: minimalCommission,
                               buySellRateMargin : buySellRateMargin
                           }
                       }, that.calcTotals);
                       console.log('fecthed');
                   })
               }
           })
       }

       isSell(){
          return (StoreUtils.getStoreState() == constants.EDIT_SELL || StoreUtils.getStoreState() == constants.EDIT_SELL_STEP2);
       }

       validated(values:any) {
           let errors: string = "";
           console.log('validation:', this.state);

           this.formData = (values as IFormData);
           if (this.formData.amount < 10) {
               errors = errors + "Amount should not be less than 10. \n";
           }

           if (this.isSell() && this.formData.amount > this.state.amountRest)
           {
               console.log('sell');
               errors = errors + "Amount should not be greater than " + this.state.amountRest + ". \n";
           } else
           if (!this.isSell() && this.state.subtotal > this.state.amountRest) {
               console.log(this.state.amountRest ,' ', this.state.rate);
               let amount:number = Math.round(this.state.amountRest / this.state.rate * 100) / 100;
               errors = errors + "Amount should not be greater than " + amount + ". \n";
           }

           console.log('errors: ',errors)

           if (errors.length > 0) {
               throw new SubmissionError({_error: errors});
               return false;
           }
           return true;
       }

       // change edit form step: 1 <-> 2
       stepAction()
       {
           let step = StoreUtils.getStoreState() == constants.EDIT_BUY  || StoreUtils.getStoreState() == constants.EDIT_SELL? 2 : 1;
           return StoreUtils.getStoreState() == constants.EDIT_BUY || StoreUtils.getStoreState() == constants.EDIT_BUY_STEP2 ?
               editBuy(this.state.currencyid, step) :
               editSell(this.state.currencyid, step)
       }

       submit(values: any) {
           // print the form values to the console
           console.log('submit');

           if ((StoreUtils.getStoreState() == constants.EDIT_BUY || StoreUtils.getStoreState() == constants.EDIT_SELL)){
               store.dispatch(this.stepAction());
               console.log(store.getState());
               this.setState({step:2});
           } else
           if (this.validated(values))
           {
               console.log('saving...', this.state, values);
               console.log(constants.TR_TRANSACTION_TYPES.BUY);
               let op:number = (StoreUtils.getStoreState() == constants.EDIT_BUY_STEP2 ? constants.TR_TRANSACTION_TYPES.BUY : constants.TR_TRANSACTION_TYPES.SELL);

               console.log('params: ', op, ', ', this.state.amount, ', ', this.state.rate, ', ',  this.state.currencyid, ', ', this.state.commissionamount);

               let data = {
                   "TransactionType": op,
                   "Amount": this.state.amount,
                   "CurrencyID": this.state.currencyid,
                   "Rate": this.state.rate,
                   "Commission": this.state.commissionamount,
                   "Note": "" // N/A
               };

               axios.post('/transaction', data, StoreUtils.authHeader())
                   .then(function (response) {
                       console.log(response);
                       store.dispatch(saveEdit()); // return to grid
                   })
                   .catch(function (error) {
                       console.log(error);
                   });
           }
       }

       calcTotals() {
           console.log('calc totals ',this.state.amount,' ', this.state.settings.commission,' ', this.state.settings.surcharge, ' ',
               this.state.settings.minimalCommission);
           let commissionAmount = Math.ceil(Math.max(parseFloat((this.state.amount * this.state.settings.commission / 100.0).toString()) + parseFloat(this.state.settings.surcharge.toString()),
               this.state.settings.minimalCommission) * 100.0) / 100.0;
           console.log(commissionAmount);

           let subtotal = this.state.rate != undefined ? Math.round(this.state.amount * this.state.rate * 100.0) / 100.0 : 0;
           console.log('subtotal:', subtotal);
           let total = Math.round((subtotal + (StoreUtils.getStoreState() == constants.EDIT_BUY ? -1 : 1) * commissionAmount) * 100) / 100;
           this.setState({subtotal:subtotal, commissionamount:commissionAmount, total:total});
       }

       amountChange(event:any) {
           console.log('amount change');
           console.log('value:', event.target.value);
           let amount: number = event.target.value;
           if (amount >= 10) {
               this.setState({amount: amount}, this.calcTotals);
           }
       }

       cancelClick(event:any) {
           console.log('cancel clicked');
           store.dispatch(cancelEdit());
       }

       backClick(event:any) {
           console.log('back clicked');
           store.dispatch(this.stepAction());
           console.log(store.getState());
           this.setState({step:1});
       }

       render() {
           console.log('rendering BuySell')
           console.log(this.props);
           console.log(this.state);
           return (<BuySellForm onSubmit={this.submit}
                                amountChange={this.amountChange}
                                cancelClick={this.cancelClick}
                                backClick={this.backClick}
                                currencycode={this.state.currencycode}
                                rate={this.state.rate}
                                amount={this.state.amount}
                                subtotal={this.state.subtotal}
                                commissionamount = {this.state.commissionamount}
                                total = {this.state.total}
                                step = {this.state.step}
                   />
           )
       }
   }

let BuySellForm = (props:any) => {
    // "step" - just passed inside to force form refresh
    const {error, handleSubmit, amountChange, cancelClick, backClick, currencycode, rate, amount, subtotal, commissionamount, total, step} = props;
    console.log(props);
    console.log(props.currencycode);
    console.log(currencycode);
    console.log(amount);
    console.log(props.amount);
    console.log("store state: ", StoreUtils.getStoreState());
    const op:string = (StoreUtils.getStoreState() == constants.EDIT_BUY || StoreUtils.getStoreState() == constants.EDIT_BUY_STEP2 ? "Buy" : "Sell");

    return (<div id="dvData2">
        <form onSubmit={handleSubmit}>
            <table>
                <tbody>
                <tr>
                    <td style={{color: "red", maxWidth: "20em"}} colSpan={2}>
                        {error && <strong>{error}</strong>}
                    </td>
                </tr>
                <tr>
                    <td colSpan={2}>
                        <h6 style={{textAlign:"center"}}> {op} {currencycode}</h6>
                    </td>
                </tr>
                <tr>
                    <td>
                        <label htmlFor="amount">Amount to {op}:</label>
                    </td>
                    <td align="right">
                        <Field
                            name="amount"
                            component="input"
                            type="number"
                            onChange={props.amountChange}
                            value={props.amount}
                            disabled={StoreUtils.getStoreState() == constants.EDIT_BUY_STEP2 || StoreUtils.getStoreState() == constants.EDIT_SELL_STEP2}
                            style={{textAlign:"right"}}
                        />
                    </td>
                </tr>
                <tr>
                    <td>
                        <label>Exchange rate:</label>
                    </td>
                    <td align="right">
                        <label>{rate}</label>
                    </td>
                </tr>
                <tr>
                    <td>
                        <label>Subtotal:</label>
                    </td>
                    <td align="right">
                        <label>{subtotal}</label>
                    </td>
                </tr>
                <tr>
                    <td>
                        <label>Commission:</label>
                    </td>
                    <td align="right">
                        <label>{commissionamount}</label>
                    </td>
                </tr>
                <tr>
                    <td colSpan={2}>
                        <div className="horizontal-rule"></div>
                    </td>
                </tr>
                <tr>
                    <td>
                        <label>Total:</label>
                    </td>
                    <td align="right">
                        <label>{total}</label>
                    </td>
                </tr>
                <tr>
                    <td colSpan={2}>
                        <div style={{float: "right"}}>
                            <button type="button" onClick={props.cancelClick} style={{margin: "5px 5px"}}>Cancel
                            </button>
                            <button type="button" onClick={props.backClick} style={{margin: "5px 5px"}}
                                    hidden={StoreUtils.getStoreState() != constants.EDIT_BUY_STEP2 && StoreUtils.getStoreState() != constants.EDIT_SELL_STEP2}>Back
                            </button>
                            <button type="submit" style={{margin: "5px 5px"}}>{op}</button>
                        </div>
                    </td>
                </tr>
                </tbody>
            </table>
        </form>
        </div>
    )
}

BuySellForm = reduxForm({
    form: 'buySell',
    initialValues:{amount:100}/*,
    destroyOnUnmount: false*/
})(BuySellForm);

export default BuySell;
