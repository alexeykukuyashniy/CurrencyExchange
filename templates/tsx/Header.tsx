import * as React from 'react';
import * as ReactDOM from 'react-dom'
import {BrowserRouter as Router, Route, NavLink} from 'react-router-dom';
import Home from './Home';
import Transactions from './Transactions';
import Admin from './Admin';
import RateService from './RateService';
import {Provider} from 'react-redux'
import store, {getStoreState, isLoggedIn} from './Store';
import CELogin from "./Login";
import { Redirect } from 'react-router';

interface IHeaderData {
    value: string;
    amount: string;
    date: string;
}

function isLoginPage(){
    return window.location.href.indexOf('login') > 0;
}

class CEHeader extends React.Component<{},{usdCash: number, usdCashStr: string, rateDate: string, minimalCurrencyRest:number}> {

    constructor(props: any) {
        super(props);
        this.state = {usdCash: 0, usdCashStr: '', rateDate: '', minimalCurrencyRest: 0};
        this.fetchData = this.fetchData.bind(this);
        this.handleStateChange = this.handleStateChange.bind(this);
        this.fetchData();
        store.subscribe(this.handleStateChange);
    }

    fetchData() {
        if (!isLoggedIn())
            return;

        let that = this;

        fetch('./headerdata'
        ).then(function (response) {
            if (response.ok) {
                console.log(response);
                let data = response.json();
                data.then(data => {
                    let d = (data as IHeaderData[])[0];
                    console.log('header data: ', data, d.amount, d.date, d.value);
                    let usdCashLocal: number = parseFloat(d.amount.replace(",", ""));

                    that.setState({
                        usdCash: usdCashLocal, usdCashStr: d.amount,
                        rateDate: d.date, minimalCurrencyRest: d.value as unknown as number
                    });
                })
            }
        })
    }

    handleStateChange() {
        console.log('Header handleStateChange: ', getStoreState(), store.getState());

        if (isLoginPage() && isLoggedIn()) {
            console.log('forceUpdate');
            this.fetchData();
        }
    }

    render() {
        console.log('this.state.ratedate ', this.state.rateDate);

        let isVisible: string = !isLoggedIn() ? "none" : "";

        if (!isLoginPage() && !isLoggedIn()) {
            window.location.href = '/login'; // redirect to login page
        } else if (!isLoginPage() && (this.state.rateDate == null || this.state.rateDate == undefined)) {
            return 'Loading...';
        } else
            return <>
                <img id="imgLogo" src="./static/images/logo.png" alt="Logo" style={{display: isVisible}}/>
                <div id="dvTitle" style={{display: isVisible}}>
                    <h2>
                        Airport Currency Exchange Office
                    </h2>
                </div>

                <Router>
                    <div id="dvLinks" style={{display: isVisible}}>
                        <ul>
                            <li>
                                <NavLink to="/" activeClassName="selected">Home</NavLink>
                            </li>
                            <li>
                                <NavLink to="/trans" activeClassName="selected">Transactions</NavLink>
                            </li>
                            <li>
                                <NavLink to="/admin" activeClassName="selected">Admin</NavLink>
                            </li>
                        </ul>
                    </div>

                    <div id='header2' style={{display: isVisible}}>
                        <span>Exchange Rates shown as per {this.state.rateDate}. You have <span
                            style={{color: this.state.usdCash <= this.state.minimalCurrencyRest ? "red" : "black"}}>{this.state.usdCash}</span> USD left.</span>
                    </div>

                    <div id="dvData">
                        {isLoginPage() && isLoggedIn() ?
                            <Redirect to="/"/> : ""
                        }
                        {!isLoginPage() && !isLoggedIn() ?
                            <Redirect to="/login"/> : ""
                        }
                        <Route path='/' component={Home}/>
                        <Route path='/trans' component={Transactions}/>
                        <Route path='/admin' component={Admin}/>
                    </div>
                </Router>
                {isLoginPage() && !isLoggedIn() ?
                    <CELogin/> : ""
                }
                {isLoggedIn() ?
                    <RateService isGetRealData={false}/> : ""
                }
            </>
    }
}

    ReactDOM.render(
        <Provider store={store}>
            <CEHeader/>
        </Provider>,
        document.getElementById('header')
    );
