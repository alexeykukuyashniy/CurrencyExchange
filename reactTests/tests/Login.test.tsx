import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount, render } from 'enzyme';
import { CELogin } from './../../templates/tsx/Login';

it('test to see if the Login renders correctly', () => {

const component = renderer.create(<CELogin />)
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
