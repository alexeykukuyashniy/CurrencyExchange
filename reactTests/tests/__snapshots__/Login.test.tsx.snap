// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test to see if the Login renders correctly 1`] = `
<div
  id="dvLogin"
>
  <form
    method="POST"
  >
    <table>
      <tbody>
        <tr>
          <td />
          <td>
            <h5>
              Enter Login
            </h5>
          </td>
        </tr>
        <tr>
          <td />
          <td>
            <span
              id="spStatus"
              style={
                Object {
                  "color": "red",
                  "marginLeft": "1em",
                }
              }
            >
              
            </span>
          </td>
        </tr>
        <tr>
          <td>
            User:
          </td>
          <td>
            <input
              id="user"
              name="user"
              onChange={[Function]}
              type="text"
              value=""
            />
          </td>
        </tr>
        <tr>
          <td>
            Password:
          </td>
          <td>
            <input
              name="pwd"
              onChange={[Function]}
              type="text"
              value=""
            />
          </td>
        </tr>
        <tr>
          <td
            colSpan={2}
            style={
              Object {
                "textAlign": "right",
              }
            }
          >
            <input
              disabled={true}
              onClick={[Function]}
              type="button"
              value="Login"
            />
          </td>
        </tr>
      </tbody>
    </table>
  </form>
</div>
`;
